### PR TITLE
Uppgift id: 1.6 | Rotera planeterna kring sin egen axel (fejka det)

### PR DESCRIPTION
**Pull request by Elin1043**
Made planets rotate around an axis. When the player is on a planet, the solar system will rotate around that planet instead. When the player leaves the planet, the solar system will reset and the planet will start rotating. There is some lag when leaving a planet and the chunksHandler is not working properly. But I am not going to fix that as it has been changed in other branches. The parameter for when the player leaves/enters a planet can be adjusted later on. I think this will work better when we have reduced lag and an atmosphere, then we can really hide the transition better.

### CODE REVIEW

File: Assets/Scripts/MarchingCubes/Planet.cs [hunk 1]
```diff
@@ -55,6 +61,8 @@ public void Initialize(Transform player, int randomSeed, bool spawn)
 
         MinMaxTerrainLevel terrainLevel = new MinMaxTerrainLevel();
 
+        rotationAxis = RandomPointOnSphereEdge(radius, rand) - Vector3.zero;
```
Comment (Manfred-Hastmark) [inline Assets/Scripts/MarchingCubes/Planet.cs:64]: Here you could maybe use the built in function rand.OnUnitSphere() or rand.OnUnitSphereUniform()
The code for this would be something like the following:

```cs
// Either
rotationAxis = rand.OnUnitSphere() * radius;
// Or
rotationAxis = rand.OnUnitSphereUniform() * radius;
```

Otherwise I think everything looks good!

**Comment by Manfred-Hastmark** [comment]
Got this interesting bug while testing your branch where I jumped in the spaceship on the inner planet, traveled to the pink planet, then clicked on F before I became a child of the pink planet, then my spaceship landed somewhere high above the solarsystem in space. I don't know if this is an issue with this branch though, could maybe be an overall bug, I'll let you decide that. Here's an image of it
image

File: Assets/Scripts/MarchingCubes/Planet.cs (hunk 1 above)
Comment (Elin1043) [reply Assets/Scripts/MarchingCubes/Planet.cs]: This is now fixed

**Comment by Elin1043** [comment]
> Got this interesting bug while testing your branch where I jumped in the spaceship on the inner planet, traveled to the pink planet, then clicked on F before I became a child of the pink planet, then my spaceship landed somewhere high above the solarsystem in space. I don't know if this is an issue with this branch though, could maybe be an overall bug, I'll let you decide that. Here's an image of it image

I have not been able to recreate this. And I don't think this is anything on my side, we should probably take a look at it later on, but probably not going to fix it in this pull request.

**Comment by Turbobus** [comment]
image

I got this while flying away from the starting planet and tried to get back. The ground just disappeared when I got close and it seemed like it took forever to load in the higher resolution planet from the beginning. This is maybe not a bug with the rotation but something has happened to the thing that decides what to load in and when.

**Comment by Turbobus** [comment]
image
While trying to reproduce the result I got above I managed to be teleported to the center of the universe inside the sun. The image is tainted because the watershader thinks there should be water there but you can see the sun marked in the hierarchy and in the scene view.

I got this by flying in and out from the starting planet over and over. Also is the rotation of the spaceship fixed? Because some times it seems to jump and it is not that it lags when the planet is changing.

**Comment by Turbobus** [comment]
image
Also managed to get inside a moon without it loading in. Don't know if this is because of rotations or not.

**Comment by Manfred-Hastmark** [comment]
> > Got this interesting bug while testing your branch where I jumped in the spaceship on the inner planet, traveled to the pink planet, then clicked on F before I became a child of the pink planet, then my spaceship landed somewhere high above the solarsystem in space. I don't know if this is an issue with this branch though, could maybe be an overall bug, I'll let you decide that. Here's an image of it image
> 
> 
> 
> I have not been able to recreate this. And I don't think this is anything on my side, we should probably take a look at it later on, but probably not going to fix it in this pull request.

Agree on that

File: Assets/Scripts/MarchingCubes/Planet.cs (hunk 1 above)
Comment (Manfred-Hastmark) [reply Assets/Scripts/MarchingCubes/Planet.cs]: 👍

**Comment by Turbobus** [comment]
image
I have the opposite problem to being able to get out when you in reality should not be able to do so. I can't get out of the spaceship at all when going to another planet (This is planet 2).

**Comment by Turbobus** [comment]
To summarize my findings: 
There are a lot of bugs when going to and from planets. I don't know if it is only the rotations fault but the thing when I happened to be teleported to the center of the universe inside the sun I think is it. The other planet stuff I don't know what is the cause but you are able to go inside many of the planets and have them be unloaded when they should not.

Also is it possible to increase the spin speed of the planets, right now they spin very slowly but I think it could be cool if it where possible to see it more?

**Comment by Elin1043** [comment]
> image
> 
> Also managed to get inside a moon without it loading in. Don't know if this is because of rotations or not. If I fly further in I can go inside the moon.
> 
> image
> 
> This also seems to be linked with the main planet unloading when coming up directly close when the highest resolution should be loaded. When going back to the main planet it unloaded when I got close as shown in my first comment.

This is on purpose. Since the moons are not working properly, I disabled being able to become a child of them. So, this will be fixed in the other task about getting the moons to be in the center of the solar system.

**Comment by Manfred-Hastmark** [comment]
> image
> 
> I got this while flying away from the starting planet and tried to get back. The ground just disappeared when I got close and it seemed like it took forever to load in the higher resolution planet from the beginning. This is maybe not a bug with the rotation but something has happened to the thing that decides what to load in and when.

I think I have fixed this problem. I had a bug in the chunkshandler

**Comment by Elin1043** [comment]
> 
This is fixed with Manfreds one line of code, it was related to the chunks.

**Comment by Elin1043** [comment]
> image While trying to reproduce the result I got above I managed to be teleported to the center of the universe inside the sun. The image is tainted because the watershader thinks there should be water there but you can see the sun marked in the hierarchy and in the scene view.
> 
> I got this by flying in and out from the starting planet over and over. Also is the rotation of the spaceship fixed? Because some times it seems to jump and it is not that it lags when the planet is changing.

I have not been able to recreate this. I flew in and out of the planet around 10 times and I had to problem. Maybe was fixed with Manfreds code? I don't know how this happened to you, but If you get it again, let me know with some reproduction steps or I don't think this is an issue right now.

Regarding the rotation of the spaceship, no it is not fixed. I am not setting the rotation of the player at all when leaving/entering a planet. The issue is related to the lag of moving the solar system and updating the chunks, you may not notice it so much but it is there. The player rotation looks like it will jump if you are moving around with the mouse as you are leaving/exiting the planets. You should find that this does not happen if you are not moving your mouse when leaving/entering. Not sure what I can do about it since forcing the player rotation to be the same before and after will also looks like it is jumping.

**Comment by Elin1043** [comment]
> image I have the opposite problem to being able to get out when you in reality should not be able to do so. I can't get out of the spaceship at all when going to another planet (This is planet 2).

This was not related to this pull request since the teleportation of the planet and setting the player as a child to it worked. The problem was that the attractor of the player was not being updated (this must have been the case for a while). You can see in the picture that the player is a child of planet 1 and the attractor of the player is still planet 0. I have fixed this now.
image

**Comment by Elin1043** [comment]
Summary:
1. Ground disappearing bug: 

-  Is fixed by Manfred. Was related to ChunksHandler.

2. Player teleporter inside the sun bug: 

- Could not recreate, may have been fixed. Please comment with reproduction steps if you get it again. Otherwise, probably not an issue for now.

3. Player can get inside the moon: 

- Intentional, the moons as we know, are not working properly. That is why I disabled being able to become a child of them for now, so we can fix them completely in the other task.

4. Player could not land on another planet bug:

- Was not related to this pull request but I fixed it anyway. The issue was that the attractor of the player for some reason was never updated, it was always set to be the first planet. Now it is being updated so that the attractor is the planet the player became a child of.

5. Rotation of spaceship:

- The rotation of the spaceship is never set by the code when entering/exiting a planet. This is related to lag when the solar system is moved and the chunks are updated all at the same time. There is a lag spike from this, not as noticeable anymore but it is still there. Not sure what I can do about it. You will see that the rotations are fine as long as you are not moving your mouse when leaving/entering a planet. I tried different methods to decrease the jumps when moving the mouse, but no results. Better to wait until we can really fix the lag or if someone has a solution to this.

6. Request of increasing rotation speed:

- The rotation speed has been increased and it is now a lot more visible.

I believe this pull request is ready to be merged if you can't recreate the bug of the player getting inside the moon or if you don't have any suggestions to the player rotation problem. But overall, the rotations and movement of the planets are working as intended, so the rotation issue is not really related to this.

**Comment by Turbobus** [comment]
I will look this through later but I will comment on the mouse/ship rotation thing. I tested it while moving the mouse and when I didn't touch it and sure the rotation errors where more common when moving the mouse because of the slight lagspike but there where times when I did not move the mouse and rotation errors still happened. 

Also what is causing the slight lagspike? I assume it is from planet having to switch meshes because there aren't too many objects to move at the same time and from Sebastians video he could do it every frame with his solar system. If it has to delete all creatures and foliage at the same time you could maybe have this happen earlier to stagger out the lag a little more (not in this pull requests).

**Comment by Elin1043** [comment]
I have not been able to get the rotation error if I'm not moving my mouse. Are you sure you did not move it? I have tested it over 20+ times and it always works if I don't touch the mouse. But again, this would be related to lag issues so I don't see why this would stop this pull request from being merged. The player controllers could probably also be updated to take time.deltatime into consideration or something like that to smooth the rotation so the jumps would be improved, but this should be done in another task.

The lag is caused by the teleportation of the planets, activation of orbits, removal of foliage and animal, switching meshes and yeah a bunch of objects being disabled/enabled at one. So yes, these things can be improved on so they don't all happen at the same time, but not in this pull request.

**Comment by Turbobus** [comment]
You are right about the rotation, that is should be done in another pull request probably connected to fixing better ship movement. I'm 95% sure that I got it to jump without touching the mouse but it is rare. The reason for taking it up in the first place was because from what I remembered you said you where going to try and fix it (at the same time as rotation) but this should not hold back rotations from getting merged. 

I will look though it later and see more about the sun bug and merge if I can't reproduce it or anything else.

**Comment by Turbobus** [comment]
Except that the main branch needs to be merged in. Are you really sure that the rotations are correct?
image

Right now if there are many moons to a planet they spawn in a disc (think Saturns rings). This disc will always be kept constant together regardless of planet rotation in the center. This is not the behavior I see when looking in the scene view currently. Here all moons have orbits that change from each other and does not follow the spawning disc relationship. What I am saying is that all moon orbits should change as one so that the moons still have the same relation to each other. (Ask if you want better explanation but I think if it like Saturns rings (which is a lot of separate objects) and that the moons are these separate objects). 

A way you can test that it is wrong currently is if you put up the rotation speed of the planet to 100 in the planet script (This speed should also be randomized so not all planets have the same rotation speed as I assume they currently has). Then the moons basically stays in the same place in the sky at all time. They slowly drift but in reality they would spin around fast around the planet.
image

Also if you think of the planets as a disc too around the sun (which the planet you are on) the other planets do not move correctly when being faked. Currently they stay with the sun as it can go in another plane (than the disc) which moves the current planet of the disc which is incorrect. By turning up the rotation speed you could also see that the sun does not work properly too as it moves to slow in the sky (nearly unchanged from before rotations). It would become day and night super fast as the planet rotates / fake rotates which it does not.

You also notice very clearly that something is of if you have such a fast rotation and going away from the planet. While at the planet it does not look like it spinns compared to the sun but when the player goes to space and the planet change, it directly start spinning att full speed.

Correct me if I'm wrong but if you have the rotation axis of a planet (which you do), instead of rotating the planet through that axis the whole solar system should rotate around that axis instead while the planet stays still? You only need to rotate everything around that axis instead of the planet?

As it currently is I will not accept this pull reqest as the rotations are not correct, this pull request has therefor not fulfilled it's purpose. Sorry :/

**Review by Turbobus — CHANGES_REQUESTED** [review]
The correct rotations needs to be fixed before I consider this pull request done. This is more commented on above this review.

Here there are smaller code/comment things that will be quick and easy to fix.

**Comment by Turbobus** [comment]
image
It was this I was talking about during our meeting. The current planet that you are on is clearly not inside the planetary disc (which it should be) which therefore has caused the planet to change orbit effectively. When going from the planet I assume that you fix this but it will cause a jump in planet position / solar system position so that the planet is back inside the planetary disc again.

I can not see how this would look normal and right in game? The sun and the whole planetary disc orbits almost vertically around the planet which causes a miss alignment that is not correct and would look like all the planets and the sun simultaneously would move up and down when stood on the planet. This is not equal to rotation.

It is possible that the solar orbit is correct (and changes correctly) (I don't know) but the other planets are not corrected for as they too should if this is the case. If the sun moves correctly the planetary disc should be rotated so that the planet you are on is in there too.

Basically, instead of the planet rotating around the planet axis, the whole solar system should rotate around that axis instead and this includes orbits and will modify the world rotation axis for the other planets too. Please ask if you don't understand what I mean by any of this!
